src/can.rs

```diff
@@ -0,0 +1,511 @@
+    fn is_remote_frame(&self) -> bool {
+        self.data().is_none()
+    }
+
+    #[inline(always)]
+    fn is_data_frame(&self) -> bool {
+        self.data().is_some()
+    }
+
+    fn id(&self) -> <CanFrame as Frame>::Id {
+        self.id
+    }
+
+    #[inline(always)]
+    fn data(&self) -> Option<&[u8]> {
+        if self.length > 0 {
+            Some(&self.data[0..self.length])
+        } else {
+            None
+        }
+    }
+}
+
+impl embedded_hal_can::Filter for CanFilter {
+    type Id = CanId;
+
+    fn from_id(id: Self::Id) -> Self {
+        CanFilter::new(CanFilterData::IdFilter(id))
+    }
+
+    fn accept_all() -> Self {
+        CanFilter::new(CanFilterData::AcceptAll)
+    }
+
+    // TODO: Constructing filters like this is fairly limiting because ideally we would have the full "filter state" available, so for non-extended filters this could be 2 masks and filters or 4 ids for id lists
+    fn from_mask(mask: u32, filter: u32) -> Self {
+        assert!(
+            mask < MAX_EXTENDED_ID,
+            "Mask cannot have bits higher than 29"
+        );
+
+        // If the mask has any significant (i.e. 1) bits in the upper bits, this filter is extended
+        if (mask & EXID_MASK) > 0 {
+            CanFilter::new(CanFilterData::ExtendedMaskFilter(mask, filter))
+        } else {
+            CanFilter::new(CanFilterData::MaskFilter(mask as u16, filter as u16))
+        }
+    }
+}
+
+impl CanFilter {
+    pub fn new(data: CanFilterData) -> CanFilter {
+        CanFilter { data, index: None }
+    }
+
+    pub fn new_with_index(index: u8, data: CanFilterData) -> CanFilter {
+        CanFilter {
+            data,
+            index: Some(index),
+        }
+    }
+}
+
+impl CanFilterData {
+    fn fr1(&self) -> u32 {
+        match self {
+            CanFilterData::AcceptAll => 0,
+            CanFilterData::ExtendedMaskFilter(filter, _) => filter << 3,
+            CanFilterData::MaskFilter(filter, mask) => {
+                let shifted_filter = ((*filter as u32) << 5) & (u16::max_value() as u32); // Only use lower 16 bits
+                let shifted_mask = ((*mask as u32) << 5) << 16;
+
+                shifted_filter | shifted_mask
+            }
+            CanFilterData::IdFilter(id) => match id {
+                CanId::BaseId(_base_id) => todo!("Shift base ID correctly!"),
+                CanId::ExtendedId(ext_id) => ext_id << 3,
+            },
+        }
+    }
+
+    fn fr2(&self) -> Option<u32> {
+        match self {
+            CanFilterData::AcceptAll => Some(0),
+            CanFilterData::ExtendedMaskFilter(_, mask) => Some(mask << 3),
+            CanFilterData::MaskFilter(_, _mask) => None, // TODO: We should be able to fill this register with a second filter/mask pair
+            CanFilterData::IdFilter(_id) => None, // TODO: This sucks, we need more info here to figure out the correct value of fr2
+        }
+    }
+}
+
+impl Can {
+    pub fn new(
+        can: stm32::CAN,
```

Comment:
Even though rust has the `new` convention for constructors, naming the constructor as the peripheral name is really common in the rust hal implementations. This is import e.g. for SPI to differentiate between `spi1, spi2, ...`. As stm32f3 devices do not have multiple can peripherals this is not as important, but maybe we should stick to it and name it `can`?